{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "##MAKE SURE FILE HAS 8k CELLS and lots of genes (either 3000 most variable or 17k)\n",
    "\n",
    "import anndata\n",
    "\n",
    "# Load the .h5ad file\n",
    "realfile = '/Users/aribahuda/Downloads/annotated_real8k.h5ad'\n",
    "activafile = '/Users/aribahuda/Downloads/annotated_activa8k.h5ad'\n",
    "real = anndata.read_h5ad(realfile)\n",
    "activa = anndata.read_h5ad(activafile)\n",
    "\n",
    "\n",
    "print(real.X.shape)\n",
    "print(activa.X.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "##Preprocessing / FILTER TO ONLY HIGHLY VARIABLE GENES (3k)\n",
    "\n",
    "import scanpy as sc\n",
    "\n",
    "sc.pp.filter_cells(real, min_genes=200)\n",
    "sc.pp.filter_genes(real, min_cells=3)\n",
    "real.var['mt'] = real.var_names.str.startswith('MT-')  # annotate the group of mitochondrial genes as 'mt'\n",
    "sc.pp.calculate_qc_metrics(real, qc_vars=['mt'], percent_top=None, log1p=False, inplace=True)\n",
    "sc.pp.normalize_total(real, target_sum=1e4)\n",
    "sc.pp.log1p(real)\n",
    "sc.pp.highly_variable_genes(real, n_top_genes=3000, flavor='cell_ranger', inplace=True)\n",
    "real = real[:, real.var.highly_variable]\n",
    "sc.pp.regress_out(real, ['total_counts', 'pct_counts_mt'])\n",
    "sc.pp.scale(real, max_value=10)\n",
    "sc.pp.neighbors(real, n_neighbors=5, n_pcs=8)\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "sc.pp.filter_cells(activa, min_genes=200)\n",
    "sc.pp.filter_genes(activa, min_cells=3)\n",
    "activa.var['mt'] = activa.var_names.str.startswith('MT-')  # annotate the group of mitochondrial genes as 'mt'\n",
    "sc.pp.calculate_qc_metrics(activa, qc_vars=['mt'], percent_top=None, log1p=False, inplace=True)\n",
    "sc.pp.normalize_total(activa, target_sum=1e4)\n",
    "sc.pp.log1p(activa)\n",
    "sc.pp.highly_variable_genes(activa, n_top_genes=3000, flavor='cell_ranger', inplace=True)\n",
    "adata = activa[:, activa.var.highly_variable]\n",
    "sc.pp.regress_out(activa, ['total_counts', 'pct_counts_mt'])\n",
    "sc.pp.scale(activa, max_value=10)\n",
    "sc.pp.neighbors(activa, n_neighbors=5, n_pcs=8)\n",
    "\n",
    "\n",
    "##Double check\n",
    "print(real.X.shape)\n",
    "print(activa.X.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "##Add Cell Type Annotations\n",
    "\n",
    "if activa.shape[0] != real.shape[0]:\n",
    "    raise ValueError(\"Number of observations (rows) in 'activa' and 'real' must be the same.\")\n",
    "\n",
    "# Set row names of 'activa' to be the same as 'real'\n",
    "activa.obs_names = real.obs_names\n",
    "\n",
    "import pandas as pd\n",
    "\n",
    "# Reading in the original dataset labels\n",
    "ann68kpbmc = pd.read_csv(\"https://raw.githubusercontent.com/10XGenomics/single-cell-3prime-paper/master/pbmc68k_analysis/68k_pbmc_barcodes_annotation.tsv\", sep='\\t')\n",
    "\n",
    "\n",
    "real_val_barcodes = real.obs_names.to_frame(index=False).rename(columns={real.obs_names.name: 'Barcodes'})['Barcodes'].tolist()\n",
    "real_validation_celltype = ann68kpbmc[ann68kpbmc['barcodes'].isin(real_val_barcodes)].copy()\n",
    "\n",
    "activa_val_barcodes = activa.obs_names.to_frame(index=False).rename(columns={activa.obs_names.name: 'Barcodes'})['Barcodes'].tolist()\n",
    "activa_validation_celltype = ann68kpbmc[ann68kpbmc['barcodes'].isin(activa_val_barcodes)].copy()\n",
    "#print(real.obs)\n",
    "\n",
    "import numpy as np\n",
    "\n",
    "# Assuming 'real' and 'activa' are your AnnData objects\n",
    "\n",
    "# Generate dataset labels for the metadata\n",
    "real_labs = np.full_like(real.obs_names, fill_value=\"Real (Test Set)\")\n",
    "activa_labs = np.full_like(activa.obs_names, fill_value=\"ACTIVA\")\n",
    "\n",
    "# Adding dataset label to metadata\n",
    "real.obs['dataset_label'] = pd.Categorical(real_labs)\n",
    "\n",
    "# Adding Celltype to metadata\n",
    "real.obs['celltype'] = pd.Categorical(real_validation_celltype['celltype'])\n",
    "\n",
    "# ACTIVA data\n",
    "\n",
    "# Adding dataset label to metadata\n",
    "activa.obs['dataset_label'] = pd.Categorical(activa_labs)\n",
    "\n",
    "# Adding Celltype to metadata\n",
    "activa.obs['celltype'] = pd.Categorical(activa_validation_celltype['celltype'])\n",
    "\n",
    "\n",
    "var_names = activa.var_names.intersection(real.var_names)\n",
    "activa = activa[:, var_names]\n",
    "real = real[:, var_names]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "##CLUSTERING + UMAP\n",
    "\n",
    "random_state = 42\n",
    "import scanpy as sc\n",
    "sc.pp.pca(activa)\n",
    "sc.pp.neighbors(activa, random_state=random_state)\n",
    "sc.tl.leiden(activa, random_state=random_state)\n",
    "sc.pl.umap(activa, color='leiden', random_state=random_state)\n",
    "\n",
    "sc.pp.pca(real)\n",
    "sc.pp.neighbors(real, random_state=random_state)\n",
    "sc.tl.leiden(real, random_state=random_state)\n",
    "sc.tl.umap(real)\n",
    "sc.pl.umap(real, color='leiden', random_state=random_state)\n",
    "\n",
    "print(real.obs['leiden'].value_counts())\n",
    "print(activa.obs['leiden'].value_counts())\n",
    "\n",
    "sc.tl.ingest(real, activa, obs='leiden')\n",
    "sc.pl.umap(real, color=['leiden', 'celltype'], wspace=0.5)\n",
    "sc.pl.umap(activa, color=['leiden', 'celltype'], wspace=0.5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "##MARKER GENE DISTRIBUTION PLOTS\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "marker_genes = ['CD79A', 'MS4A1', 'CD8A', 'CD8B', 'LYZ', 'CD14', 'LGALS3', 'S100A8',\n",
    "                'GNLY', 'NKG7', 'FCGR3A', 'MS4A7', 'FCER1A', 'CST3', 'PPBP']\n",
    "\n",
    "\n",
    "#sc.pl.dotplot(real, marker_genes, 'leiden', color_map='Blues')\n",
    "#sc.pl.dotplot(activa, marker_genes, 'leiden')\n",
    "\n",
    "#sc.pl.dotplot(real, marker_genes, 'celltype', vmin=0, vmax=1)\n",
    "#sc.pl.dotplot(activa, marker_genes, 'celltype', vmin=0, vmax=1)\n",
    "\n",
    "\n",
    "expression_values = real[:, marker_genes].X\n",
    "\n",
    "plt.figure(figsize=(10, 6))\n",
    "plt.boxplot(expression_values, vert=False, labels=marker_genes)\n",
    "plt.xlabel(\"Expression\")\n",
    "plt.ylabel(\"Genes\")\n",
    "plt.xlim(-1,11)\n",
    "plt.title(\"Marker Genes - Real\")\n",
    "plt.show()\n",
    "\n",
    "expression_values = activa[:, marker_genes].X\n",
    "\n",
    "# Step 5: Plot the gene expression values using a horizontal box and whisker plot\n",
    "plt.figure(figsize=(10, 6))\n",
    "plt.boxplot(expression_values, vert=False, labels=marker_genes)\n",
    "plt.xlabel(\"Expression\")\n",
    "plt.xlim(-1,11)\n",
    "plt.ylabel(\"Genes\")\n",
    "plt.title(\"Marker Genes - ACTIVA\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "##PERCENT SHARED DEGS\n",
    "\n",
    "import scanpy as sc\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "# Assuming 'real' and 'activa' are your Scanpy AnnData objects\n",
    "\n",
    "# Perform differential expression analysis for both datasets\n",
    "sc.tl.rank_genes_groups(real, 'leiden', method='wilcoxon')\n",
    "sc.tl.rank_genes_groups(activa, 'leiden', method='wilcoxon')\n",
    "\n",
    "# Extract top 20 differentially expressed genes for each dataset and each cluster\n",
    "top_genes_real = real.uns['rank_genes_groups']['names']\n",
    "top_genes_activa = activa.uns['rank_genes_groups']['names']\n",
    "\n",
    "# Get the list of leiden clusters\n",
    "leiden_clusters = real.obs['leiden'].cat.categories\n",
    "\n",
    "# Initialize lists to store shared DEGs and total DEGs within each cluster\n",
    "shared_degs_percentages = []\n",
    "total_degs = []\n",
    "\n",
    "# Iterate over each Leiden cluster\n",
    "for cluster in leiden_clusters:\n",
    "    # Get the top 20 DEGs for this cluster in each dataset\n",
    "    top_degs_real = set(top_genes_real[cluster][:100])\n",
    "    top_degs_activa = set(top_genes_activa[cluster][:100])\n",
    "    \n",
    "    # Calculate the intersection (shared DEGs) and union (total DEGs) of the two sets\n",
    "    shared_degs = top_degs_real.intersection(top_degs_activa)\n",
    "    total_degs_cluster = len(top_degs_real) + len(top_degs_activa) - len(shared_degs)\n",
    "    \n",
    "    # Calculate the percentage of shared DEGs within this cluster\n",
    "    percentage_shared_degs = (len(shared_degs) / total_degs_cluster) * 100\n",
    "    \n",
    "    # Append the results to the lists\n",
    "    shared_degs_percentages.append(percentage_shared_degs)\n",
    "    total_degs.append(total_degs_cluster)\n",
    "\n",
    "# Plotting\n",
    "x = np.arange(len(leiden_clusters))  # x-axis for each cluster\n",
    "plt.bar(x, shared_degs_percentages)\n",
    "plt.xlabel('Leiden Cluster')\n",
    "plt.ylabel('Percentage of Shared DEGs (%)')\n",
    "plt.title('Percentage of Shared DEGs within Leiden Clusters')\n",
    "plt.xticks(x, labels=leiden_clusters)\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "##DEG HEATMAP\n",
    "\n",
    "import scanpy as sc\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "\n",
    "# Assuming 'real' and 'activa' are your Scanpy AnnData objects\n",
    "\n",
    "# Perform differential expression analysis for both datasets\n",
    "sc.tl.rank_genes_groups(real, 'leiden', method='wilcoxon')\n",
    "sc.tl.rank_genes_groups(activa, 'leiden', method='wilcoxon')\n",
    "\n",
    "# Get the ordered list of Leiden clusters\n",
    "leiden_clusters_order = real.obs['leiden'].cat.categories.astype(int)\n",
    "\n",
    "# Plot rank genes groups matrixplot for 'real' dataset\n",
    "sc.pl.rank_genes_groups_matrixplot(real, n_genes=3,  groupby='leiden', swap_axes=True, cmap='viridis', standard_scale='var', show=None, dendrogram=False)\n",
    "plt.gca().set_title('Real Dataset')\n",
    "plt.xticks(ticks=range(len(leiden_clusters_order)), labels=leiden_clusters_order)\n",
    "\n",
    "# Plot rank genes groups matrixplot for 'activa' dataset\n",
    "sc.pl.rank_genes_groups_matrixplot(activa, n_genes=3, groupby='leiden', swap_axes=True, cmap='viridis', standard_scale='var', show=None, dendrogram=False)\n",
    "plt.gca().set_title('Activa Dataset')\n",
    "plt.xticks(ticks=range(len(leiden_clusters_order)), labels=leiden_clusters_order)\n",
    "\n",
    "plt.tight_layout()\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "##DISTRIBUTION OF VARIANCE\n",
    "\n",
    "gene_expression_variance = real.X.var(axis=0)\n",
    "\n",
    "# Step 2: Plot the distribution of gene expression variance\n",
    "plt.figure(figsize=(8, 6))\n",
    "plt.hist(gene_expression_variance, bins=50, color='skyblue', edgecolor='black')\n",
    "plt.xlabel('Variance of Gene Expression')\n",
    "plt.ylabel('Frequency')\n",
    "plt.title('Distribution of Gene Expression Variance')\n",
    "plt.show()\n",
    "\n",
    "gene_expression_variance = activa.X.var(axis=0)\n",
    "\n",
    "# Step 2: Plot the distribution of gene expression variance\n",
    "plt.figure(figsize=(8, 6))\n",
    "plt.hist(gene_expression_variance, bins=50, color='skyblue', edgecolor='black')\n",
    "plt.xlabel('Variance of Gene Expression')\n",
    "plt.ylabel('Frequency')\n",
    "plt.title('Distribution of Gene Expression Variance')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "language_info": {
   "name": "python"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
